In [34]:
# =================================================================
# Hücre 1: Kurulum ve Veri Yükleme
# =================================================================
import pandas as pd
import numpy as np

# Yolunu senin projene göre ayarla
DATA_PATH = "../data/raw/gps_locations.csv"

df = pd.read_csv(DATA_PATH)

print("[OK] Veri yüklendi:", DATA_PATH)
print("Satır:", len(df), "| Sütun:", df.shape[1])


[OK] Veri yüklendi: ../data/raw/gps_locations.csv
Satır: 120000 | Sütun: 26


In [35]:
# Hücre 2: Veri Ön İncelemesi

print("\n--- İlk 5 Satır ---")
display(df.head())

print("\n--- Veri Bilgisi ---")
print(df.info())

print("\n--- Eksik Değerler ---")
print(df.isna().sum())

print("\n--- Özet İstatistikler ---")
display(df.describe())



--- İlk 5 Satır ---


trip_id  driver_id  vehicle_id            timestamp   latitude   longitude  \
0        1        101        1001  2023-01-01 00:00:00  38.916143  -80.345269   
1        2        105        2002  2023-01-01 00:00:01  37.011830  -89.079516   
2        3        103        2002  2023-01-01 00:00:02  33.784009  -99.103643   
3        4        102        2002  2023-01-01 00:00:03  45.314835 -102.046210   
4        5        101        2002  2023-01-01 00:00:04  30.473386  -92.362577   

       speed  acceleration  steering_angle     heading  ...  \
0  39.484646      3.612832               6  212.555994  ...   
1   2.795422     -0.593295             -21  134.541146  ...   
2  13.842558     -0.209264              26  116.452839  ...   
3  69.121833     -0.413943               6   26.970247  ...   
4  13.061111      5.774122              -1  333.804023  ...   

   weather_conditions  road_type  traffic_condition  stop_events  \
0               Sunny      Urban              Light            0   
1               Sunny    Highway              Light            0   
2              Cloudy      Urban           Moderate            0   
3               Sunny    Highway           Moderate            3   
4               Sunny      Rural           Moderate            0   

   geofencing_violation  anomalous_event route_anomaly route_deviation_score  \
0                     0                0             0              0.176974   
1                     0                1             0              0.229316   
2                     0                0             0              0.473809   
3                     0                0             0              0.368063   
4                     0                0             0              0.145270   

  acceleration_variation  behavioral_consistency_index  
0               0.715284                      0.468189  
1               0.487761                      0.234935  
2               0.881141                      0.192588  
3               0.937918                      0.363707  
4               0.702187                      0.178748  

[5 rows x 26 columns]


--- Veri Bilgisi ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120000 entries, 0 to 119999
Data columns (total 26 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   trip_id                       120000 non-null  int64  
 1   driver_id                     120000 non-null  int64  
 2   vehicle_id                    120000 non-null  int64  
 3   timestamp                     120000 non-null  object 
 4   latitude                      120000 non-null  float64
 5   longitude                     120000 non-null  float64
 6   speed                         120000 non-null  float64
 7   acceleration                  120000 non-null  float64
 8   steering_angle                120000 non-null  int64  
 9   heading                       120000 non-null  float64
 10  trip_duration                 120000 non-null  float64
 11  trip_distance                 120000 non-null  float64
 12  fuel_consumption      

trip_id      driver_id     vehicle_id       latitude  \
count  120000.000000  120000.000000  120000.000000  120000.000000   
mean    60000.500000     102.098875    2106.479375      39.990066   
std     34641.160489       1.371367    1379.539988       5.002472   
min         1.000000     101.000000    1001.000000      17.930570   
25%     30000.750000     101.000000    1001.000000      36.623453   
50%     60000.500000     102.000000    2002.000000      39.987503   
75%     90000.250000     103.000000    3003.000000      43.357958   
max    120000.000000     105.000000    5005.000000      61.096832   

           longitude          speed   acceleration  steering_angle  \
count  120000.000000  120000.000000  120000.000000   120000.000000   
mean      -90.002296      29.971987       1.003904        0.082442   
std        10.026103      30.211853       1.998322       14.870230   
min      -133.129729       0.000157      -0.999910      -45.000000   
25%       -96.799608       8.615250      -0.419030      -10.000000   
50%       -89.969652      20.633107       0.388639        0.000000   
75%       -83.241541      41.439832       1.777851       10.000000   
max       -46.981523     365.764111      22.032532       45.000000   

             heading  trip_duration  ...            rpm    brake_usage  \
count  120000.000000  120000.000000  ...  120000.000000  120000.000000   
mean      179.923038    3593.505077  ...    1991.533620       4.995767   
std       103.833915    3595.552160  ...    1498.933743       2.240727   
min         0.000915       0.001821  ...     500.015651       0.000000   
25%        89.972635    1032.752956  ...     927.900033       3.000000   
50%       179.859502    2484.164733  ...    1535.258915       5.000000   
75%       269.673973    4975.872175  ...    2563.837376       6.000000   
max       359.996851   46866.782872  ...   18879.432692      19.000000   

       lane_deviation    stop_events  geofencing_violation  anomalous_event  \
count   120000.000000  120000.000000         120000.000000    120000.000000   
mean         0.285753       0.996800              0.050242         0.100167   
std          0.159819       1.001481              0.218444         0.300223   
min          0.000354       0.000000              0.000000         0.000000   
25%          0.161382       0.000000              0.000000         0.000000   
50%          0.264586       1.000000              0.000000         0.000000   
75%          0.389144       2.000000              0.000000         0.000000   
max          0.932762       8.000000              1.000000         1.000000   

       route_anomaly  route_deviation_score  acceleration_variation  \
count  120000.000000          120000.000000           120000.000000   
mean        0.099708               0.285187                0.714811   
std         0.299612               0.159840                0.159045   
min         0.000000               0.000305                0.075593   
25%         0.000000               0.160065                0.611532   
50%         0.000000               0.263214                0.735854   
75%         0.000000               0.389066                0.838992   
max         1.000000               0.948786                0.999540   

       behavioral_consistency_index  
count                 120000.000000  
mean                       0.285855  
std                        0.160088  
min                        0.000295  
25%                        0.160986  
50%                        0.264427  
75%                        0.389694  
max                        0.942612  

[8 rows x 22 columns]

### Sonuç & Öneri
- Veri seti başarıyla yüklendi (120.000 satır, 26 sütun).  
- Eksik değer bulunmamaktadır.  
- Veri tipleri büyük ölçüde uygun görünüyor; yalnızca `timestamp` sütunu `object` tipinde ve dönüştürülmesi gerekiyor.  
- Aykırı değer olarak yüksek hız (365 km/h) gözlemlendi, ileride işaretlenmesi yararlı olacaktır.  

➡️ Sonraki adım: `timestamp` sütununun `datetime` tipine dönüştürülmesi ve zaman bazlı özniteliklerin türetilmesi (gün, saat, vb.).  


In [36]:
# =========================================================
# H3: Zaman Damgası Dönüşümü + Bayraklar (tek sabit ile)
# Not: Temizlik YAPMAZ; sadece işaretler (flag).
# =========================================================

# -- Tek yerden yönetilen hız eşiği (km/s)
if "SPEED_CEILING_KMH" not in globals():
    SPEED_CEILING_KMH = 200  # ihtiyaca göre 180/220 olarak güncellenebilir

# 1) timestamp -> datetime
df["timestamp"] = pd.to_datetime(df["timestamp"], errors="coerce")

# 2) zaman türevleri
df["date"] = df["timestamp"].dt.date
df["hour"] = df["timestamp"].dt.hour.astype("Int64")  # geçici Int64, H4'te istersek int32'ye çeviririz
df["day_of_week"] = df["timestamp"].dt.day_name()

# 3) kontrol bayrakları
df["is_ts_missing"]   = df["timestamp"].isna().astype(int)
df["is_gps_valid"]    = (
    df["latitude"].between(-90, 90, inclusive="both")
    & df["longitude"].between(-180, 180, inclusive="both")
).astype(int)
df["is_speed_suspect"] = (df["speed"] >= SPEED_CEILING_KMH).astype(int)

# 4) rapor
n = len(df)
bad_ts = int(df["is_ts_missing"].sum())
bad_gps = int((df["is_gps_valid"] == 0).sum())
sus_spd = int(df["is_speed_suspect"].sum())

print("=== Dönüşüm & Bayrak Özeti (H3) ===")
print(f"Toplam satır: {n:,}")
print(f"Geçersiz timestamp (NaT): {bad_ts}")
print(f"Geçersiz GPS (aralık dışı): {bad_gps}")
print(f"Şüpheli hız (≥ {SPEED_CEILING_KMH} km/h): {sus_spd}")

print("\n[Zaman aralığı]")
print("Başlangıç:", df["timestamp"].min())
print("Bitiş   :", df["timestamp"].max())


=== Dönüşüm & Bayrak Özeti (H3) ===
Toplam satır: 120,000
Geçersiz timestamp (NaT): 0
Geçersiz GPS (aralık dışı): 0
Şüpheli hız (≥ 200 km/h): 183

[Zaman aralığı]
Başlangıç: 2023-01-01 00:00:00
Bitiş   : 2023-01-02 09:19:59


**H3 – Kısa, net özet**

* **Veri Durumu:** Veri, zaman ve konum açısından sağlam. Eksik veya hatalı zaman damgası (NaT) ve aralık dışı GPS verisi bulunmuyor.
* **Anormallik:** Yalnızca hız tarafında 183 aşım var (≥ 200 km/h, ~%0.15). Bu anormallikler, bir sonraki adımda temizlenecek şekilde bayraklandı.
* **Aksiyon:** Veri, bir sonraki aşama olan yıkıcı temizlik (satır silme) için uygun zeminde. Bu aşamada sadece bayraklama yapıldığı için, ham verinin bir kopyası korunmaktadır.

In [37]:
# =========================================================
# H4: Temizleme Kuralları (bayraklara göre drop + rapor)
# - H3'te üretilen bayrakları kullanır
# - Hız eşiği H3 ile AYNI sabitten alınır: SPEED_CEILING_KMH
# - Drop öncesi en hızlı 10 kaydı loglar (denetim izi)
# =========================================================

import pandas as pd

# -- Temizlik kuralları
ENFORCE_TS_VALID   = True   # NaT olan timestamp satırlarını at
ENFORCE_GPS_RANGE  = True   # -90/90, -180/180 dışını at
ENFORCE_SPEED_CEIL = True   # is_speed_suspect==1 olanı at

# -- Denetim izi: drop ÖNCESİ en hızlı 10 kayıt (gözlem için)
print("\n[Ön Kontrol] En hızlı 10 kayıt (drop öncesi):")
cols_preview = ["timestamp", "vehicle_id", "driver_id", "speed", "latitude", "longitude"]
print(df.sort_values("speed", ascending=False)[cols_preview].head(10))

# -- Silme maskesi
to_drop_mask = pd.Series(False, index=df.index)
if ENFORCE_TS_VALID:
    to_drop_mask |= (df["is_ts_missing"] == 1)
if ENFORCE_GPS_RANGE:
    to_drop_mask |= (df["is_gps_valid"] == 0)
if ENFORCE_SPEED_CEIL:
    to_drop_mask |= (df["is_speed_suspect"] == 1)

# -- Neden bazında sayımlar (rapor)
reason_counts = {
    "invalid_timestamp": int((df["is_ts_missing"] == 1).sum()),
    "invalid_gps":       int((df["is_gps_valid"] == 0).sum()),
    f"speed_≥{SPEED_CEILING_KMH}": int((df["is_speed_suspect"] == 1).sum()),
}

n_before = len(df)
n_drop   = int(to_drop_mask.sum())
n_after  = n_before - n_drop

print("\n=== Temizleme Özeti (H4) ===")
print(f"Önceki satır sayısı : {n_before:,}")
print(pd.DataFrame.from_dict(reason_counts, orient="index", columns=["silinecek_satır"]))
print(f"Toplam silinecek    : {n_drop:,}")

# -- Gerçek drop
if n_drop > 0:
    df = df.loc[~to_drop_mask].copy()
    df.reset_index(drop=True, inplace=True)

print(f"Sonraki satır sayısı : {len(df):,}")

# -- Tür optimizasyonu: kategorikler
print("\n--- Kategorik Değişkenler Dönüştürülüyor ---")
categorical_features = [
    "driver_id", "vehicle_id",
    "weather_conditions", "road_type", "traffic_condition",
    "day_of_week"
]
for col in categorical_features:
    if col in df.columns:
        df[col] = df[col].astype("category")

# hour'ı kompakt tiplendir (opsiyonel)
if "hour" in df.columns:
    df["hour"] = df["hour"].astype("int32")

# -- Hızlı kontrol
print("\n[Kontrol] Kısaltılmış info:")
print(df[["timestamp","latitude","longitude","speed"]].info())



[Ön Kontrol] En hızlı 10 kayıt (drop öncesi):
                 timestamp  vehicle_id  driver_id       speed   latitude  \
115297 2023-01-02 08:01:37        1001        104  365.764111  47.546566   
23883  2023-01-01 06:38:03        3003        104  343.730216  44.157061   
94366  2023-01-02 02:12:46        1001        101  337.340720  46.119289   
78697  2023-01-01 21:51:37        2002        104  332.391756  42.545817   
104861 2023-01-02 05:07:41        1001        105  327.469383  42.045349   
65560  2023-01-01 18:12:40        5005        104  311.801512  42.824321   
14152  2023-01-01 03:55:52        1001        101  307.209545  37.062115   
31535  2023-01-01 08:45:35        3003        101  306.300331  32.725816   
19515  2023-01-01 05:25:15        1001        102  305.928998  33.935666   
76791  2023-01-01 21:19:51        1001        101  296.061678  34.991385   

        longitude  
115297 -96.125793  
23883  -93.954593  
94366  -94.176272  
78697  -79.566127  
104861 -68.53584

**H4 – Kısa, net özet**

* **Temizleme:** Hız eşiğini aşan 183 satır (`is_speed_suspect == 1`) temizlendi. Zaman ve GPS kaynaklı kayıp yaşanmadı.
* **Veri Boyutu:** Toplam satır sayısı 120,000'den 119,817'ye düştü.
* **Optimizasyon:** `driver_id`, `vehicle_id`, `weather_conditions`, `road_type`, `traffic_condition` gibi kimlik ve bağlamsal sütunlar, bellek kullanımını azaltmak ve grup işlemlerini hızlandırmak amacıyla `category` tipine dönüştürüldü.
* **Durum:** Veri, sonraki adım olan **Yolculuk Segmentasyonu** ve **Özellik Mühendisliği** için temiz ve optimize edilmiş durumda.

In [38]:
# =========================================================
# Hücre 5: trip_id Tekillik Kontrolü (EDA → Preprocessing köprüsü)
# Veri Seti Kontrolü: trip_id Tekil Mi?
# =========================================================

import pandas as pd

# df'in güncel durumunu kontrol et
total_rows = len(df)
unique_trip_ids = df['trip_id'].nunique()

print(f"Toplam satır sayısı: {total_rows:,}")
print(f"Tekil trip_id sayısı: {unique_trip_ids:,}")

if total_rows == unique_trip_ids:
    print("\n[Uyarı] Her bir GPS kaydı, ayrı bir yolculuk olarak tanımlanmış.")
    print("Yolculuk segmentasyonunu yeniden yapmamız gerekiyor.")
else:
    print("\n[Bilgi] Yolculuklar doğru gruplandırılmış.")

Toplam satır sayısı: 119,817
Tekil trip_id sayısı: 119,817

[Uyarı] Her bir GPS kaydı, ayrı bir yolculuk olarak tanımlanmış.
Yolculuk segmentasyonunu yeniden yapmamız gerekiyor.


**Değerlendirme – Trip ID Kontrolü**  
- Toplam satır: 119,817 | Tekil trip_id: 119,817  
- Sonuç: Her satır için farklı `trip_id` atanmış. Yani veri seti, her GPS noktasını ayrı yolculuk gibi tanımlıyor.  
- Etki: Mevcut `trip_id` sütunu yolculuk segmentasyonu için kullanılamaz. Bu nedenle kendi segmentasyonumuzu zaman farkı kuralı ile yeniden tanımlamamız gerekiyor.  

In [39]:
# =========================================================
# H6: Yolculuk Segmentasyonu + Temiz Veriyi Kaydetme
# Amacı: Yolculuk segmentlerini oluşturmak ve temiz df'yi kaydetmek
# =========================================================

import os
import datetime as dt
import pandas as pd

# Güvenlik: df var mı?
if "df" not in globals():
    raise RuntimeError("df bulunamadı. Önceki hücreleri çalıştırdığınızdan emin olun.")

print("=== Yolculuk Segmentasyonu (H6) – araç bazında ===")

# --- Parametreler
# 5 dakika (300 saniye) ve 10 dakika (600 saniye) yaygın olarak kullanılır.
# 5 dakika, dur-kalk veya kısa molalar için daha uygun olabilir.
GAP_SEC = 300       # 5 dakika
MIN_ROWS_PER_SEG = 2 # en az 2 satır şartı (tek nokta segmentleri eleyelim)

# 1) Güvenli sıralama: önce vehicle_id, sonra timestamp
df.sort_values(by=['vehicle_id', 'timestamp'], inplace=True)

# 2) Araç bazında ardışık zaman farkı (saniye)
df['time_diff_sec'] = df.groupby('vehicle_id')['timestamp'].diff().dt.total_seconds().fillna(0)

# 3) Yeni segment bayrağı: araç değişimi veya zaman boşluğu > GAP_SEC
df['new_seg'] = ((df['vehicle_id'] != df['vehicle_id'].shift()) | (df['time_diff_sec'] > GAP_SEC)).astype(int)

# 4) Araç içi kümülatif toplama ile yerel segment id'si
df['trip_seg_local'] = df.groupby('vehicle_id')['new_seg'].cumsum()

# 5) Global benzersiz segment id
df["trip_seg_id"] = pd.factorize(df['vehicle_id'].astype(str) + '_' + df['trip_seg_local'].astype(str))[0] + 1

# Segment boyutlarına göre filtreleme
seg_counts = df.groupby('trip_seg_id').size()
valid_segments = seg_counts[seg_counts >= MIN_ROWS_PER_SEG].index
df = df[df['trip_seg_id'].isin(valid_segments)].copy()

print(f"Toplam segment sayısı (min {MIN_ROWS_PER_SEG} satır): {df['trip_seg_id'].nunique():,}")
print(f"Segmentasyondan sonraki satır sayısı: {len(df):,}")

print("\n--- Temiz Seti Kaydetme ---")
# --- Kayıt yolları
CLEAN_DIR = "../data/clean"
BASE_NAME = "gps_locations_clean.csv"
TS = dt.datetime.now().strftime("%Y%m%d_%H%M%S")
VERSIONED_NAME = f"gps_locations_clean_{TS}.csv"
os.makedirs(CLEAN_DIR, exist_ok=True)
base_path = os.path.join(CLEAN_DIR, BASE_NAME)
ver_path  = os.path.join(CLEAN_DIR, VERSIONED_NAME)

# --- Kaydet
df.to_csv(base_path, index=False)
df.to_csv(ver_path,  index=False)

print(f"[OK] Ana dosya  : {base_path}")
print(f"[OK] Sürüm dosya: {ver_path}")
print(f"Satır/Sütun     : {df.shape[0]:,} x {df.shape[1]}")

=== Yolculuk Segmentasyonu (H6) – araç bazında ===
Toplam segment sayısı (min 2 satır): 5
Segmentasyondan sonraki satır sayısı: 119,817

--- Temiz Seti Kaydetme ---


/var/folders/ng/1b3xfg1d3gbbbqgbn43lfd880000gn/T/ipykernel_84003/3248343457.py:26: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df['time_diff_sec'] = df.groupby('vehicle_id')['timestamp'].diff().dt.total_seconds().fillna(0)
/var/folders/ng/1b3xfg1d3gbbbqgbn43lfd880000gn/T/ipykernel_84003/3248343457.py:32: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df['trip_seg_local'] = df.groupby('vehicle_id')['new_seg'].cumsum()


[OK] Ana dosya  : ../data/clean/gps_locations_clean.csv
[OK] Sürüm dosya: ../data/clean/gps_locations_clean_20250821_011937.csv
Satır/Sütun     : 119,817 x 36


### Hücre 6: Yolculuk Segmentasyonunu Oluşturma ve Temiz Veriyi Kaydetme

Bu hücre, veri ön işleme sürecinin en kritik adımlarından biridir. 
Ham veri setindeki her GPS kaydının ayrı bir `trip_id`'ye sahip olduğunu `Hücre 5`'te tespit ettik. 
Bu durum, yolculuk bazında özellikler oluşturmamızı engelliyordu.

Bu hücre iki temel görevi yerine getirir:
1.  **Yolculuk Segmentasyonu:** Art arda gelen GPS kayıtları arasındaki zaman farkına bakarak mantıksal yolculuk segmentleri (`trip_seg_id`) oluşturur. Bu, araç bazında gerçek sürüş gruplarını tanımlar.
2.  **Temiz Veriyi Kaydetme:** Oluşturulan `trip_seg_id` sütununu içeren nihai, temizlenmiş veri setini `../data/clean/gps_locations_clean.csv` dosyasına kaydeder. Bu dosya, bir sonraki aşama olan `03_feature_engineering.ipynb` not defterinde kullanılacaktır.

Bu hücrenin doğru çalışması, tüm projenin başarısı için zorunludur.